Error compiling and serving scully app

# Error compiling and serving scully app

I created this issue because I'm having problems creating my first scully app. After following the steps in the Getting Started tutorial I'm getting unexpected responses in the console. Everything goes well until I get to the serve/build step.

I created this [minimal reproduction of the issue](https://github.com/danmt/scully-building-issue).

If I execute `npm run scully` I get as an output:

```

> my-scully-app@0.0.0 scully /home/danmt/Blogging/Scully/my-scully-app
> scully

Cleaned up /home/danmt/Blogging/Scully/my-scully-app/dist/static folder.
 ☺   new Angular build imported
Background servers already running.
Could not connect to server
npm ERR! code ELIFECYCLE
npm ERR! errno 15
npm ERR! my-scully-app@0.0.0 scully: `scully`
npm ERR! Exit status 15
npm ERR!
npm ERR! Failed at the my-scully-app@0.0.0 scully script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/danmt/.npm/_logs/2020-01-22T23_35_18_956Z-debug.log

```

If I execute `npm run scully:serve` I get as an output:

```

> my-scully-app@0.0.0 scully:serve /home/danmt/Blogging/Scully/my-scully-app
> scully serve

starting static server
Adding all extra routes (1)
Could not start Scully serve TypeError: Cannot read property 'length' of undefined

```

I updated the `scully.config.js` file to this:

```javascript
exports.config = {
  projectRoot: "./src/app",
  outDir: "./dist/static",
  routes: {
    "/blog/:slug": {
      type: "contentFolder",
      slug: {
        folder: "./blog"
      }
    }
  },
  extraRoutes: [""]
};
```

@danmt I solved the issue with the extra routes.
It looks to me you had a stale _old_ version running. 
```
Cleaned up /home/danmt/Blogging/Scully/my-scully-app/dist/static folder.
 ☺   new Angular build imported
Background servers already running.
Could not connect to server
npm ERR! code ELIFECYCLE
```
The message `Background servers already running.` tells me you already had a version running.
If you see a message like that, you should run `npx scully killServer` so you are sure you are on the new version.
We will take some precations so we will make sure your server ends on installing a new version.

--- a/scully.config.js
+++ b/scully.config.js
@@ -9,6 +9,7 @@ exports.config = {
   /** outDir is where the static distribution files end up */
   outDir: './dist/static',
   // hostName: '0.0.0.0',
+  extraRoutes: [''],
   routes: {
     '/demo/:id': {
       type: 'extra',

--- a/scully/routerPlugins/traverseAppRoutesPlugin.ts
+++ b/scully/routerPlugins/traverseAppRoutesPlugin.ts
@@ -1,7 +1,7 @@
 import {parseAngularRoutes} from 'guess-parser';
-import {scullyConfig} from '../utils/config';
-import {logError, logWarn, log, green, yellow} from '../utils/log';
 import * as yargs from 'yargs';
+import {scullyConfig} from '../utils/config';
+import {green, logError, logWarn, yellow} from '../utils/log';
 
 const {sge} = yargs
   .boolean('sge')
@@ -47,53 +47,56 @@ ${green(`By adding '' to the extraRoutes array in the scully.config option, you 
 };
 
 export async function addExtraRoutes(): Promise<string[]> {
-  const extraRoutes: any[] = scullyConfig.extraRoutes;
+  const isPromise = (x: any) => x && x.then !== undefined && typeof x.then === 'function';
+  const extraRoutes: string | (string | Promise<string | string[]>)[] | Promise<string | string[]> =
+    scullyConfig.extraRoutes;
   if (!extraRoutes) {
-    return Promise.resolve([]);
+    return [];
   }
-
-  if (!Array.isArray(extraRoutes)) {
-    logWarn(`ExtraRoutes must be provided as an array. Current type: ${typeof extraRoutes}`);
-    return Promise.resolve([]);
-  } else {
-    log(`Adding all extra routes (${extraRoutes.length})`);
-    const extraRoutePromises = extraRoutes.map(extraRoute => {
-      if (!extraRoute) {
-        return Promise.resolve();
+  if (typeof extraRoutes === 'string') {
+    /** convert a simple string to an array */
+    return [extraRoutes];
+  }
+  const workList: (string | Promise<string | string[]>)[] = [];
+  if (isPromise(extraRoutes)) {
+    const outerResult = await extraRoutes;
+    if (typeof outerResult === 'string') {
+      /** ok, we got a promise<string> return the result */
+      return [outerResult];
+    }
+    workList.concat(outerResult);
+  } else if (Array.isArray(extraRoutes)) {
+    extraRoutes.forEach(r => {
+      if (workList.includes(r)) {
+        /** don't add duplicates */
+        return;
       }
-      // It is a promise already
-      if (extraRoute.then && {}.toString.call(extraRoute.then) === '[object Function]') {
-        return extraRoute;
-      }
-
-      // Turn into promise<string>
-      if (typeof extraRoute === 'string') {
-        return Promise.resolve(extraRoute);
-      }
-
-      logWarn(
-        `The extraRoute ${JSON.stringify(
-          extraRoute
-        )} needs to be either a string or a Promise<string|string[]>. Excluding for now. `
-      );
-      // Turn into promise<undefined>
-      return Promise.resolve();
+      workList.push(r);
     });
-    const extraRouteValues = await Promise.all(extraRoutePromises);
-    extraRouteValues.reduce((acc, val) => {
-      // Remove empties and just return acc
-      if (!val) {
-        return acc;
-      }
-
-      // Spread acc and arrays together
-      if (Array.isArray(val)) {
-        return [...acc, ...val];
-      }
-
-      // Append values into acc
-      return [...acc, val];
-    }, []);
-    return extraRouteValues;
+  } else {
+    logWarn(`ExtraRoutes must be provided as an string array. Current type: ${typeof extraRoutes}`);
+    return [];
   }
+
+  const result: string[] = [];
+  for (const route of workList) {
+    /** note, this ignores non-string/non-promise things in array */
+    if (typeof route === 'string') {
+      result.push(route);
+    }
+    if (isPromise(route)) {
+      const x = await route;
+      if (typeof x === 'string') {
+        result.push(x);
+      }
+      if (Array.isArray(x)) {
+        x.forEach(s => {
+          if (typeof s === 'string') {
+            result.push(s);
+          }
+        });
+      }
+    }
+  }
+  return result;
 }

--- a/scully/utils/interfacesandenums.ts
+++ b/scully/utils/interfacesandenums.ts
@@ -20,7 +20,7 @@ export interface ScullyConfig {
   /** routes that needs additional processing have their configuration in here */
   routes: RouteConfig;
   /** routes that are in the application but have no route in the router */
-  extraRoutes?: string[];
+  extraRoutes?: (string | Promise<string[] | string>)[];
   /** Port-number where the original application is served */
   appPort: number;
   /** port-number where the Scully generated files are available */

